plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3ExchangeStorage.java

```diff
@@ -0,0 +1,686 @@
+
+                ObjectMetadata metadata = new ObjectMetadata();
+                metadata.setContentLength(bufferSize);
+                metadata.setContentMD5(getMd5AsBase64(buffer, 0, bufferSize));
+
+                PutObjectRequest request = new PutObjectRequest(bucketName, key, in, metadata);
+                sseCustomerKey.ifPresent(request::withSSECustomerKey);
+
+                try {
+                    s3.putObject(request);
+                    return;
+                }
+                catch (AmazonServiceException e) {
+                    throw new IOException(e);
+                }
+            }
+
+            // The multipart upload API only accept the last part to be less than 5MB
+            if (bufferSize == buffer.length || (finished && bufferSize > 0)) {
+                byte[] data = buffer;
+                int length = bufferSize;
+                this.buffer = new byte[buffer.length];
+                bufferSize = 0;
+
+                inProgressUploadFuture = uploadExecutor.submit(() -> uploadPage(data, length));
+            }
+        }
+
+        private UploadPartResult uploadPage(byte[] data, int length)
+        {
+            if (uploadId.isEmpty()) {
+                uploadId = Optional.of(uploadIdFactory.get());
+            }
+
+            currentPartNumber++;
+            UploadPartRequest uploadRequest = new UploadPartRequest()
+                    .withBucketName(bucketName)
+                    .withKey(key)
+                    .withUploadId(uploadId.get())
+                    .withPartNumber(currentPartNumber)
+                    .withInputStream(new ByteArrayInputStream(data, 0, length))
+                    .withPartSize(length)
+                    .withMD5Digest(getMd5AsBase64(data, 0, length));
+            sseCustomerKey.ifPresent(uploadRequest::withSSECustomerKey);
+
+            UploadPartResult partResult = s3.uploadPart(uploadRequest);
+            parts.add(partResult);
+            return partResult;
+        }
+
+        private static String getMd5AsBase64(byte[] data, int offset, int length)
+        {
+            @SuppressWarnings("deprecation")
+            byte[] md5 = md5().hashBytes(data, offset, length).asBytes();
+            return Base64.getEncoder().encodeToString(md5);
+        }
+
+        private void waitForPreviousUploadFinish()
+                throws IOException
+        {
+            if (inProgressUploadFuture == null) {
+                return;
+            }
+
+            try {
+                inProgressUploadFuture.get();
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new InterruptedIOException();
+            }
+            catch (ExecutionException e) {
+                throw new IOException("Streaming upload failed", e);
+            }
+        }
+
+        private void finishUpload(String uploadId)
```

Comment:
maybe call this eagerly in `flushBuffer(true)`, not only in close

Reply:
I feel the current way is clearer and easier to understand?